syntax = "proto3";

package archivematica.ccp.admin.v1beta1;

import "archivematica/ccp/admin/v1beta1/i18n.proto";
import "buf/validate/validate.proto";
import "google/protobuf/timestamp.proto";

message Package {
  // Identifier of the package (UUIDv4).
  string id = 1 [(buf.validate.field).string.uuid = true];

  // Name of the transfer.
  string name = 2 [(buf.validate.field).string.min_len = 1];

  // Type of the transfer.
  TransferType type = 3;

  // Status of the package.
  PackageStatus status = 4;

  // Creation timestamp.
  google.protobuf.Timestamp created_at = 5;

  // Current directory.
  string directory = 6;

  // Access system identifier.
  string access_system_id = 7;

  // Whether the package has been hidden by the user.
  bool hidden = 8;

  // List of associated jobs. It may not always be populated.
  repeated Job job = 9;
}

message Job {
  // Identifier of the job (UUIDv4).
  string id = 1 [(buf.validate.field).string.uuid = true];

  // Identifier of the package (UUIDv4).
  string package_id = 2 [(buf.validate.field).string.uuid = true];

  PackageType package_type = 3 [
    (buf.validate.field).required = true,
    (buf.validate.field).enum = {
      defined_only: true,
      in: [
        1, // "unitTransfer"
        2, // "unitSIP"
        4 // "unitDIP"
      ]
    }
  ];

  string link_id = 4 [(buf.validate.field).string.uuid = true];

  JobStatus status = 5;

  google.protobuf.Timestamp created_at = 6;

  string group = 7;

  string link_description = 8;

  bool hidden = 9;

  string directory = 10;

  // Only populated if requested.
  // TODO: this is not efficient but it's currently needed.
  Decision decision = 11;
}

message Decision {
  // Identifier of the decision (UUIDv4).
  string id = 1 [(buf.validate.field).string.uuid = true];

  // Name of the decision (prompt).
  string name = 2 [(buf.validate.field).string.min_len = 1];

  // Identifier of the package (UUIDv4).
  string package_id = 3 [(buf.validate.field).string.uuid = true];

  string package_path = 4;

  string package_type = 5; // "Transfer", "SIP", "DIP".

  // Identifier of the job (UUIDv4).
  string job_id = 6 [(buf.validate.field).string.uuid = true];

  // Ordered list of choices. The position can be used to resolve a decision.
  repeated Choice choice = 7 [(buf.validate.field).repeated = {min_items: 1}];
}

message Choice {
  // Identifier of the choice based on its position in the ordered choice list.
  int32 id = 1 [(buf.validate.field).int32 = {gte: 0}];

  // Label or description of the choice.
  string label = 2 [(buf.validate.field).string.min_len = 1];
}

message ProcessingConfigField {
  string id = 1;
  string name = 2;
  I18n label = 3;
  repeated ProcessingConfigFieldChoice choice = 4;
}

message ProcessingConfigFieldChoice {
  string value = 1;
  I18n label = 2;
  repeated ProcessingConfigFieldChoiceAppliesTo applies_to = 3;
}

message ProcessingConfigFieldChoiceAppliesTo {
  string link_id = 1;
  string value = 2; // E.g. URI, replacement ID, chainID...
  I18n label = 3;
}

// Different types of transfers.
enum TransferType {
  TRANSFER_TYPE_UNSPECIFIED = 0;
  TRANSFER_TYPE_STANDARD = 1;
  TRANSFER_TYPE_ZIP_FILE = 2;
  TRANSFER_TYPE_UNZIPPED_BAG = 3;
  TRANSFER_TYPE_ZIPPED_BAG = 4;
  TRANSFER_TYPE_DSPACE = 5;
  TRANSFER_TYPE_MAILDIR = 6;
  TRANSFER_TYPE_TRIM = 7;
  TRANSFER_TYPE_DATAVERSE = 8;
}

enum PackageType {
  PACKAGE_TYPE_UNSPECIFIED = 0;
  PACKAGE_TYPE_TRANSFER = 1;
  PACKAGE_TYPE_SIP = 2;
  PACKAGE_TYPE_AIP = 3;
  PACKAGE_TYPE_DIP = 4;
}

enum PackageStatus {
  PACKAGE_STATUS_UNSPECIFIED = 0;
  PACKAGE_STATUS_PROCESSING = 1;
  PACKAGE_STATUS_DONE = 2;
  PACKAGE_STATUS_COMPLETED_SUCCESSFULLY = 3;
  PACKAGE_STATUS_FAILED = 4;

  // This is not found in the database but we can infer it at runtime.
  PACKAGE_STATUS_AWAITING_DECISION = 5;
}

enum JobStatus {
  JOB_STATUS_UNSPECIFIED = 0;
  JOB_STATUS_AWAITING_DECISION = 1;
  JOB_STATUS_COMPLETED_SUCCESSFULLY = 2;
  JOB_STATUS_EXECUTING_COMMANDS = 3;
  JOB_STATUS_FAILED = 4;
}
